### DISCUSSION TITLE
loadable atom doesn't have type for returned value `data` and `error` properties

### DISCUSSION THREAD

**Discussion by trongthanh** [pmndrs/jotai#2731] [Bug report] [answered]
### Bug Description

Jotai 2.9.3

The returned value from a `useAtom(loadableAtom)` only has type definition for `state`. And no type inference for the `data` property based on the async atom. Besides, `error` doesn't exist as a property.

![image](https://github.com/user-attachments/assets/67916252-2210-4dda-915f-0b06caf88ef7)

### Reproduction Link

https://stackblitz.com/edit/jotai-gmwmbz?file=src%2FApp.tsx

**Comment by dai-shi** [comment]
You can check the state and then access it.
https://stackblitz.com/edit/jotai-gmwmbz-lfvpq2?file=src%2FApp.tsx

**Reply by thanhtran-ct** [reply]
Of course I can write check routine to make sure `data` is defined. My bug (or feature request) is that can the loadable atom infer the returned type of the wrapped async atom and report the type of `data` property as union with `undefined`? For example, in this case, `data` property should exist with type `string | undefined`?

For the `error` property, it is fine as is. User can check and cast the type case by case

**Reply by dai-shi** [reply, marked as answer]
https://github.com/pmndrs/jotai/blob/9b835851a57b5e7333aa4c52a7f1833bea24b8e9/src/vanilla/utils/loadable.ts#L17-L20 is the type we are using. We generally avoid adding `| undefined` in all Jotai code base. So, your request is not acceptable.

If you want to loosen the type, you can work it around with type assertion:
https://stackblitz.com/edit/jotai-gmwmbz-rbdn9a?file=src%2FApp.tsx
It's not very cute though.

**Reply by trongthanh** [reply]
Agree, type assertion at userland is counterproductive of type convenience. However, as I'm writing a wrapper around `loadable` so this is acceptable for now.
